### PR TITLE
Convert the scripts to TypeScript

### PR DESCRIPTION
**Pull request by j-f1**
Fixes https://github.com/desktop/desktop/issues/48.

TODO:

- [x] `app-info.js`
- [x] `git-info.js`
- [x] `package-info.js`
- [x] `dev-commands-global.js`
- [x] `webpack.*.js`
- ~[`debug`](https://github.com/desktop/desktop/blob/master/script/debug) (imports [`express`](https://www.npmjs.com/package/@types/express), [`webpack`](https://www.npmjs.com/package/@types/webpack), [`webpack-dev-middleware`](https://www.npmjs.com/package/@types/webpack-dev-middleware), and [`webpack-hot-middleware`](https://www.npmjs.com/package/@types/webpack-hot-middleware)) (maybe merge with `start`?)~
- [x] [`dist-info.js`](https://github.com/desktop/desktop/blob/master/script/dist-info.js)
- [x] [`license-overrides.js`](https://github.com/desktop/desktop/blob/master/script/license-overrides.js) (depended upon by `build.ts`) (indirectly depends on `legal-eagle` which ~has no `@types` package defined~ has TYPES at [`@types/legal-eagle`](https://www.npmjs.com/package/@types/legal-eagle))
- [x] [`publish`](https://github.com/desktop/desktop/blob/master/script/publish) (imports [`aws-sdk`](https://www.npmjs.com/package/aws-sdk) and [`request`](https://www.npmjs.com/package/@types/request))
- [x] [`run.js`](https://github.com/desktop/desktop/blob/master/script/run.js) (depended upon by [`start`](https://github.com/desktop/desktop/blob/master/script/start))
- [x] [`start`](https://github.com/desktop/desktop/blob/master/script/start) (imports [`express`](https://www.npmjs.com/package/@types/express), [`webpack`](https://www.npmjs.com/package/@types/webpack), [`webpack-dev-middleware`](https://www.npmjs.com/package/@types/webpack-dev-middleware), and [`webpack-hot-middleware`](https://www.npmjs.com/package/@types/webpack-hot-middleware))

### CODE REVIEW

File: script/publish.ts [hunk 1]
```diff
@@ -124,36 +128,39 @@ function upload(assetName, assetPath) {
...
+      (error: Error, data: AWS.S3.ManagedUpload.SendData) => {
+        if (error) {
+          reject(error)
+        } else {
+          const stats = Fs.statSync(assetPath)
+          const hash = Crypto.createHash('sha1')
+          const input = Fs.createReadStream(assetPath)
+
+          hash.on('finish', () => {
+            const sha = hash.read().toString('hex')
```
Comment (j-f1) [inline script/publish.ts]: TS *really* doesn’t like this. Should this be `hash.digest().toString('hex')`?

Comment (shiftkey) [reply script/publish.ts]: `digest` computes based on the current content, but I still think we want to `read` here and process some data. We've assumed here that `read()` returns a `Buffer`, but we don't check for `string` here. That's what TS is complaining about.

TL;DR: if we get a string here, I'm not sure what we should do about it. Let's throw here and figure it out later:

```ts
hash.on('finish', () => {
  const result = hash.read()
  if (result instanceof Buffer) {
    const sha = result.toString('hex')
    resolve({ name: assetName, url, size: stats['size'], sha })
  } else {
    reject('hash.read() returned a string?!?!?!?')
  }
})
```

Comment (j-f1) [reply script/publish.ts]: Is the latest commit better?

Comment (shiftkey) [reply script/publish.ts]: Yeah, that seems better.

File: package.json [hunk 2]
```diff
@@ -31,7 +31,7 @@
     "eslint": "ts-node script/eslint.ts",
     "eslint:fix": "yarn eslint --fix",
     "eslint-check": "eslint --print-config .eslintrc.* | eslint-config-prettier-check",
-    "publish": "node script/publish",
+    "publish": "ts-node script/publish",
```
Comment (shiftkey) [inline package.json]: Don't forget the `.ts` on the end here

File: script/debug.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
+'use strict'
```
Comment (shiftkey) [inline script/debug.ts]: Is this line still necessary?

**Comment by j-f1** [comment]
Depends on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/22722.

**Comment by j-f1** [comment]
🔷

**Comment by j-f1** [comment]
Can someone restart the Travis CI build?

**Comment by j-f1** [comment]
🐢

**Comment by iAmWillShepherd** [comment]
@j-f1 looks like you need to do some conflict resolution! 👍

**Comment by j-f1** [comment]
@iAmWillShepherd Conflicts are gone 🔥

**Comment by iAmWillShepherd** [comment]
@j-f1 when I try to build the project, I receive the following.

**Comment by j-f1** [comment]
Did https://github.com/desktop/desktop/pull/3675/commits/4a79b8ddb20d8250142ae8e0428e733e9f47e060 fix the issue?

**Comment by iAmWillShepherd** [comment]
Sorry @j-f1, I did not realize you had committed a fix. 

👀 now.

**Comment by j-f1** [comment]
🚋

**Comment by shiftkey** [comment]
Just moving this into 1.2 as it touches a lot of infrastructure code and needs a methodical review.

**Comment by sale7-778** [comment]
[GlobalErrors.log](https://github.com/desktop/desktop/files/1759772/GlobalErrors.log)

File: package.json [hunk 4]
```diff
@@ -5,35 +5,35 @@
...
-    "compile:script": "tsc -p script",
+    "compile:script": "tsc -p script/tsconfig.json",
     "lint": "yarn lint:src && yarn lint:prettier",
     "lint:fix": "yarn lint:src:fix && yarn lint:prettier --fix",
     "lint:prettier": "ts-node script/prettier.ts",
     "lint:src": "yarn tslint && yarn eslint-check && yarn eslint",
     "lint:src:fix": "yarn tslint --fix && yarn eslint --fix",
     "tslint": "tslint ./{script,tslint-rules}/*.ts ./app/{src,typings,test}/**/*.{ts,tsx}",
-    "eslint": "ts-node script/eslint.ts",
+    "eslint": "ts-node -P script/tsconfig.json script/eslint.ts",
```
[the remark this answers is not in the document]
Comment (j-f1) [reply package.json]: 2.8 added conditional types

File: script/run.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    'Contents',
+    'MacOS',
+    `${productName}`
+  )
+} else if (process.platform === 'win32') {
+  binaryPath = join(distPath, `${productName}.exe`)
+} else if (process.platform === 'linux') {
+  binaryPath = join(distPath, productName)
+} else if (
+  process.platform === 'aix' ||
```
[the remark this answers is not in the document]
Comment (j-f1) [reply script/run.ts]: Sure, but I wanted to do an exhaustive platform check here.

**Comment by j-f1** [comment]
Bah, pressed the wrong button.

I’m on vacation right now 🌴, so I’ll have some better replies for you sometime next week.

**Comment by shiftkey** [comment]
@j-f1 all good, happy to pick this up again when you're back

**Comment by shiftkey** [comment]
Dropping this from `1.2.0` because I think we've got enough in-flight work to review to as well as upcoming feature work.

File: package.json [hunk 6]
```diff
@@ -5,23 +5,23 @@
...
-    "compile:dev": "cross-env NODE_ENV=development parallel-webpack --config app/webpack.development.js",
-    "compile:prod": "cross-env NODE_ENV=production parallel-webpack --config app/webpack.production.js",
-    "build:dev": "yarn compile:dev && cross-env NODE_ENV=development ts-node script/build.ts",
-    "build:prod": "yarn compile:prod && cross-env NODE_ENV=production ts-node script/build.ts",
+    "start": "cross-env NODE_ENV=development ts-node -P script script/start.ts",
+    "start:prod": "cross-env NODE_ENV=production ts-node -P script script/start.ts",
+    "debug": "cross-env NODE_ENV=development ts-node -P script script/debug.ts",
+    "compile:dev": "cross-env NODE_ENV=development parallel-webpack --config app/webpack.development.ts",
+    "compile:prod": "cross-env NODE_ENV=production parallel-webpack --config app/webpack.production.ts",
+    "build:dev": "yarn compile:dev && cross-env NODE_ENV=development ts-node -P script script/build.ts",
```
[the remark this answers is not in the document]
Comment (j-f1) [reply package.json]: Looks like `ts-node`, a separate project from TS, dropped support for passing a directory here in v5.

[the remark this answers is not in the document]
Comment (j-f1) [reply package.json]: (`ts-node`’s GitHub project is [here](https://github.com/TypeStrong/ts-node/))

**Comment by shiftkey** [comment]
Looks like this PR is updating itself to yarn 1.5.1, and this error is occurring:

```
 warning Lockfile has incorrect entry for "@types/react@^16.0.40". Ignoring it.
```

I can't trigger this on `master` locally, but I've pushed f4fc41e96 to get this back and building.

File: app/package-info.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+const appPackage: Record<string, string> = require('./package.json')
```
[the remark this answers is not in the document]
Comment (j-f1) [reply app/package-info.ts]: Not as far as I can see.

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (shiftkey) [reply package.json]: This was fixed in #4478 which I've merged into this branch.

**Comment by j-f1** [comment]
Is there any way to see why the VSTS build failed without getting an MS Live account?

**Comment by shiftkey** [comment]
@j-f1 that's actually failing because building PRs from forks is [currently blocked for security reasons](https://docs.microsoft.com/en-us/vsts/build-release/actions/ci-build-github?view=vsts#building-pull-requests-from-repository-forks). I've reached out to the team to ask the status of that.

**Comment by j-f1** [comment]
🍵

File: app/webpack.common.ts [hunk 8]
```diff
@@ -38,6 +34,9 @@ const commonConfig = {
             options: {
               useBabel: true,
               useCache: true,
+              compilerOptions: {
+                noEmit: false, // override from tsconfig.json
```
Comment (shiftkey) [inline app/webpack.common.ts]: @j-f1 I've dropped this override in 9810d3d as I couldn't find any problems with either `yarn && yarn build:dev && yarn start` or `yarn && yarn build:prod && yarn start:prod` the resulting bundles. Let me know if there's some context here that I've missed.

**Comment by shiftkey** [comment]
@desktop/core @desktop/comrades once we've gotten 1.2.0 out and things are stable I'd like to get this in and test the last part (publishing test builds to ensure that last part of the pipeline is fine). Aside from that I think everything here is ready for review!

**Comment by shiftkey** [comment]
https://github.com/desktop/desktop/pull/3675/commits/a02d8c4f7f87c1340ceea10c87e870f1ab311896 now merges the Webpack 4 changes into this PR, and there's a couple of type errors that are outstanding now - almost ready for another round of review!

**Comment by shiftkey** [comment]
Alright, I'm dropping myself as a reviewer because I've been heavily involved in this, but I think this is close to merging in. Could someone else give this a pass and a :thumbsup:?

**Comment by shiftkey** [comment]
Friendly bump on this @desktop/core - would love to figure out if we can land this soon

File: script/start.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
+import * as express from 'express'
+import * as webpack from 'webpack'
+import * as devMiddleware from 'webpack-dev-middleware'
+import * as hotMiddleware from 'webpack-hot-middleware'
+
+import { forceUnwrap as u } from '../app/src/lib/fatal-error'
+
+import configs = require('../app/webpack.development')
```
Comment (iAmWillShepherd) [inline script/start.ts]: ❓ Why do we use `require` here?

Comment (j-f1) [reply script/start.ts]: I’m using `import =` since that file uses `export =`.

Comment (shiftkey) [reply script/start.ts]: I think the shape of the webpack config files doesn't really give us much to work with here

File: script/publish.ts [hunk 10]
```diff
@@ -22,31 +17,49 @@ if (!releaseSHA) {
...
+import { execSync } from 'child_process'
+import * as AWS from 'aws-sdk'
+import * as Crypto from 'crypto'
+import * as request from 'request'
 
 console.log('Packaging…')
-cp.execSync('yarn package')
+execSync('yarn package')
+
+const getSha = () => {
```
Comment (iAmWillShepherd) [inline script/publish.ts]: This could be normal function as well.

File: script/publish.ts [hunk 11]
```diff
@@ -22,31 +17,49 @@ if (!releaseSHA) {
 const currentTipSHA = gitInfo.getSHA()
 if (
   !currentTipSHA ||
-  !currentTipSHA.toUpperCase().startsWith(releaseSHA.toUpperCase())
+  !currentTipSHA.toUpperCase().startsWith(releaseSHA!.toUpperCase())
```
Comment (iAmWillShepherd) [inline script/publish.ts]: `releaseSHA!` can cause an error if it's `null` or `undefined`.

File: script/publish.ts [hunk 12]
```diff
@@ -22,31 +17,49 @@ if (!releaseSHA) {
...
-  sha = process.env.APPVEYOR_REPO_COMMIT
+  throw new Error(
+    `Unable to get the SHA for the current platform. Check the vendor docs for the desired environment variables.`
+  )
 }
 
-sha = sha.substr(0, 8)
+const sha = getSha().substr(0, 8)
+
+const getSecret = () => {
```
[the remark this answers is not in the document]
Comment (j-f1) [reply script/publish.ts]: I’m pretty sure that it automagically figures out the right name for the function, but I’ll definitely change these.

File: script/start.ts (hunk 9 above)
Comment (j-f1) [reply script/start.ts]: I remember a little better now — We could use `import * as configs`, but the file exports an array, but when you use `import *`, you get an object.

File: script/publish.ts (hunk 11 above)
Comment (j-f1) [reply script/publish.ts]: I exit the script in the previous `if` statement if it’s `null` or `undefined`, but TS isn’t smart enough to pick up on that yet.

**Comment by shiftkey** [comment]
Thanks @j-f1 for this work!
